File: src/main/java/main/commands/slashcommands/commands/audio/StopCommand.java
package main.commands.slashcommands.commands.audio;

import com.sedmelluq.discord.lavaplayer.track.AudioTrack;
import main.audiohandlers.RobertifyAudioManager;
import main.commands.prefixcommands.CommandContext;
import main.commands.prefixcommands.ICommand;
import main.utils.RobertifyEmbedUtils;
import main.utils.component.interactions.AbstractSlashCommand;
import main.utils.json.requestchannel.RequestChannelConfig;
import main.utils.json.logs.LogType;
import main.utils.json.logs.LogUtils;
import main.utils.locale.RobertifyLocaleMessage;
import net.dv8tion.jda.api.EmbedBuilder;
import net.dv8tion.jda.api.entities.*;
import net.dv8tion.jda.api.events.interaction.command.SlashCommandInteractionEvent;
import net.dv8tion.jda.internal.utils.tuple.Pair;
import org.jetbrains.annotations.NotNull;

import javax.script.ScriptException;

public class StopCommand extends AbstractSlashCommand implements ICommand {
    @Override
    public void handle(CommandContext ctx) throws ScriptException {
        final Message msg = ctx.getMessage();

        msg.replyEmbeds(handleStop(ctx.getMember()).build()).queue();
    }

    public EmbedBuilder handleStop(Member stopper) {
        Guild guild = stopper.getGuild();
        final var selfVoiceState = guild.getSelfMember().getVoiceState();
        final var memberVoiceState = stopper.getVoiceState();

        if (!selfVoiceState.inAudioChannel())
            return RobertifyEmbedUtils.embedMessage(guild, RobertifyLocaleMessage.GeneralMessages.NOTHING_PLAYING);

        if (!memberVoiceState.inAudioChannel())
            return RobertifyEmbedUtils.embedMessage(guild, RobertifyLocaleMessage.GeneralMessages.USER_VOICE_CHANNEL_NEEDED);


        if (!memberVoiceState.getChannel().equals(selfVoiceState.getChannel()))
            return RobertifyEmbedUtils.embedMessage(guild, RobertifyLocaleMessage.GeneralMessages.SAME_VOICE_CHANNEL_LOC, Pair.of("{channel}", selfVoiceState.getChannel().getAsMention()));;

        final var musicManager = RobertifyAudioManager.getInstance().getMusicManager(guild);
        final var audioPlayer = musicManager.getPlayer();
        final var scheduler = musicManager.getScheduler();
        final AudioTrack track = audioPlayer.getPlayingTrack();

        if (track == null)
            return RobertifyEmbedUtils.embedMessage(guild, RobertifyLocaleMessage.GeneralMessages.NOTHING_PLAYING);

        scheduler.repeating = false;
        scheduler.playlistRepeating = false;
        audioPlayer.stopTrack();
        scheduler.queue.clear();
        scheduler.getPastQueue().clear();

        if (audioPlayer.isPaused())
            audioPlayer.setPaused(false);

        if (new RequestChannelConfig(guild).isChannelSet())
            new RequestChannelConfig(guild).updateMessage();

        new LogUtils(guild).sendLog(LogType.PLAYER_STOP, RobertifyLocaleMessage.StopMessages.STOPPED_LOG, Pair.of("{user}", stopper.getAsMention()));

        scheduler.scheduleDisconnect(true);
        return RobertifyEmbedUtils.embedMessage(guild, RobertifyLocaleMessage.StopMessages.STOPPED);
    }

    @Override
    public String getName() {
        return "stop";
    }

    @Override
    public String getHelp(String prefix) {
        return "Forces the bot to stop playing music and clear the queue" +
                " if already playing music.";
    }

    @Override
    protected void buildCommand() {
        setCommand(
                getBuilder()
                        .setName("stop")
                        .setDescription("Stop the music and clear the queue")
                        .setPossibleDJCommand()
                        .build()
        );
    }

    @Override
    public String getHelp() {
        return "Forces the bot to stop playing music and clear the queue" +
                " if already playing music.";
    }

    @Override
    public void onSlashCommandInteraction(@NotNull SlashCommandInteractionEvent event) {
        if (!checks(event)) return;

        event.replyEmbeds(handleStop(event.getMember())
                .build())
                .queue();
    }
}

File: src/main/kotlin/main/commands/slashcommands/audio/StopCommand.kt
package main.commands.slashcommands.audio

import dev.minn.jda.ktx.util.SLF4J
import main.audiohandlers.RobertifyAudioManager
import main.audiohandlers.utils.title
import main.utils.RobertifyEmbedUtils
import main.utils.RobertifyEmbedUtils.Companion.sendEmbed
import main.utils.component.interactions.slashcommand.AbstractSlashCommand
import main.utils.component.interactions.slashcommand.models.SlashCommand
import main.utils.json.logs.LogType
import main.utils.json.logs.LogUtilsKt
import main.utils.locale.messages.GeneralMessages
import main.utils.locale.messages.StopMessages
import net.dv8tion.jda.api.entities.Member
import net.dv8tion.jda.api.entities.MessageEmbed
import net.dv8tion.jda.api.events.interaction.command.SlashCommandInteractionEvent
import kotlin.time.Duration.Companion.seconds

class StopCommand : AbstractSlashCommand(
    SlashCommand(
        name = "stop",
        description = "Stop the music and clear the queue."
    )
) {

    companion object {
        private val logger by SLF4J
    }

    override suspend fun handle(event: SlashCommandInteractionEvent) {
        event.deferReply().queue()
        event.hook.sendEmbed(event.guild) { handleStop(event.member!!) }.queue()
    }

    fun handleStop(stopper: Member): MessageEmbed {
        val guild = stopper.guild
        val selfVoiceState = guild.selfMember.voiceState!!
        val memberVoiceState = stopper.voiceState!!
        val musicManager = RobertifyAudioManager[guild]
        val player = musicManager.player
        val scheduler = musicManager.scheduler
        val track = player?.track

        return when {
            !selfVoiceState.inAudioChannel() -> RobertifyEmbedUtils.embedMessage(
                guild,
                GeneralMessages.NOTHING_PLAYING
            ).build()

            !memberVoiceState.inAudioChannel() -> RobertifyEmbedUtils.embedMessage(
                guild,
                GeneralMessages.USER_VOICE_CHANNEL_NEEDED
            ).build()

            memberVoiceState.channel!!.id != selfVoiceState.channel!!.id -> RobertifyEmbedUtils.embedMessage(
                guild,
                GeneralMessages.SAME_VOICE_CHANNEL_LOC,
                Pair("{channel}", selfVoiceState.channel!!.asMention)
            ).build()

            track == null -> RobertifyEmbedUtils.embedMessage(
                guild,
                GeneralMessages.NOTHING_PLAYING
            ).build()

            else -> {
                musicManager.clear()
                player.stopTrack().subscribe()
                logger.debug("Stopped track. Playing track: ${player.track?.title ?: "none"}")

                LogUtilsKt(guild).sendLog(
                    LogType.PLAYER_STOP,
                    StopMessages.STOPPED_LOG,
                    Pair("{user}", stopper.asMention)
                )

                scheduler.scheduleDisconnect(announceMsg = true)
                return RobertifyEmbedUtils.embedMessage(guild, StopMessages.STOPPED).build()
            }
        }
    }

    override val help: String
        get() = "Forces the bot to stop playing music and clear the queue if aslready playing music."
}